libsweep/serial.h

```diff
@@ -7,24 +7,29 @@
  */
 
 #include "sweep.h"
+#include "Error.h"
 
-#include <stdint.h>
+#include <memory>
 
 namespace sweep {
 namespace serial {
 
-typedef struct device* device_s;
-typedef struct error* error_s;
-
-const char* error_message(error_s error);
-void error_destruct(error_s error);
-
-device_s device_construct(const char* port, int32_t bitrate, error_s* error);
-void device_destruct(device_s serial);
-
-void device_read(device_s serial, void* to, int32_t len, error_s* error);
-void device_write(device_s serial, const void* from, int32_t len, error_s* error);
-void device_flush(device_s serial, error_s* error);
+struct Error : public ErrorBase {
+	using ErrorBase::ErrorBase;
+};
+
+class Device {
+public:
+	Device(const char* port, int32_t bitrate);
+	Device(Device&& other) = default;
```

Comment:
I probably just forgot the move assignment operator. The pimple idiom was used to keep the header free of platform specific code (and in particular includes) just like the original. but other than that there is no reason why the state couldn't be stored in normal member variables (guarded by an appropriated `#ifdef`).